src/Datadog.Trace.ClrProfiler.Managed/AutoInstrumentation/Kafka/KafkaProduceSyncDeliveryHandlerIntegration.cs

```diff
@@ -0,0 +1,182 @@
+        /// <param name="handler">The delivery handler instance</param>
+        /// <returns>Calltarget state value</returns>
+        public static CallTargetState OnMethodBegin<TTarget, TKey, TValue, TActionOfDeliveryReport>(TTarget instance, string topic, TKey key, TValue value, TActionOfDeliveryReport handler)
+        {
+            if (handler is null)
+            {
+                // Handled in KafkaProduceSyncIntegration
+                return CallTargetState.GetDefault();
+            }
+
+            try
+            {
+                var span = Tracer.Instance?.ActiveScope?.Span;
+                if (span is null)
+                {
+                    Logger.Warning("Unexpected null span for Kafka Producer with delivery handler");
+                    return CallTargetState.GetDefault();
+                }
+
+                var newAction = CachedWrapperDelegate<TActionOfDeliveryReport>.CreateWrapper(handler, span);
+
+                Action<ITypedDeliveryHandlerShimAction> updateHandlerAction = inst =>
+                {
+                    try
+                    {
+                        inst.Handler = newAction;
+                    }
+                    catch (Exception ex)
+                    {
+                        Logger.Error(ex, "There was an error updating the delivery report handler to ");
+                        // Not ideal to close the span here immediately, but as we can't trace the result,
+                        // we don't really have a choice
+                        span.Finish();
+                    }
+                };
+
+                // store the call to update the handler variable as state
+                // so we update it at the _end_ of the constructor
+                return new CallTargetState(scope: null, state: updateHandlerAction);
+            }
+            catch (Exception ex)
+            {
+                Logger.Error(ex, "Error creating wrapped delegate for delivery report");
+                return CallTargetState.GetDefault();
+            }
+        }
+
+        /// <summary>
+        /// OnMethodEnd callback
+        /// </summary>
+        /// <typeparam name="TTarget">Type of the target</typeparam>
+        /// <param name="instance">Instance value, aka `this` of the instrumented method.</param>
+        /// <param name="exception">Exception instance in case the original code threw an exception.</param>
+        /// <param name="state">Calltarget state value</param>
+        /// <returns>A response value, in an async scenario will be T of Task of T</returns>
+        public static CallTargetReturn OnMethodEnd<TTarget>(TTarget instance, Exception exception, CallTargetState state)
+        {
+            if (state.State is Action<ITypedDeliveryHandlerShimAction> updateHandlerAction
+             && instance.TryDuckCast<ITypedDeliveryHandlerShimAction>(out var shim))
+            {
+                updateHandlerAction.Invoke(shim);
```

Comment:
Can you create a try/catch block to wrap this `Invoke` call instead of inside of the delegate? That way, we can remove the the use of the `Span` argument inside the delegate and we can get rid of that closure

Reply:
Done in https://github.com/DataDog/dd-trace-dotnet/pull/1444/commits/67ead42e5d0b9db676a513c47530817c3281cc4a